chain/chain/src/types.rs

```diff
@@ -227,6 +227,26 @@ pub trait RuntimeAdapter: Send + Sync {
         state_root: MerkleHash,
         payload: Vec<u8>,
     ) -> Result<(), Box<dyn std::error::Error>>;
+
+    /// Build receipts hashes.
+    fn build_receipts_hashes(
+        &self,
+        receipts: &Vec<ReceiptTransaction>,
+    ) -> Result<Vec<CryptoHash>, Error> {
+        let mut receipts_hashes = vec![];
+        for shard_id in 0..self.num_shards() {
+            // importance to save the same order while filtering
+            let shard_receipts: Vec<ReceiptTransaction> = receipts
+                .iter()
+                .filter(|&receipt| {
+                    self.account_id_to_shard_id(&receipt.receiver) == shard_id
+                })
+                .cloned()
+                .collect();
+            receipts_hashes.push(hash_struct(&shard_receipts));
```

Comment:
`ReceiptTransaction` has a `nonce` that uniquely identifies a receipt. Instead of hashing it every time, we can just use that for computing merkle proofs.

Reply:
Here's an example of attack: I'm malicious, and I take the receipts I requested to send, then replace all the receipts but save original nonce, and send for the nodes who takes OnePart's only.
If you take OnePart for a shard, you calculate hash of nonces and get a faked proof that received receipts are valid, and then apply faked receipts - that will be found in distant future when someone will check all the chunks in the shard.
Does it make sense? Maybe there are some strong checks for nonces in receipts which protect from such kind of attack?

Reply:
Good point. I don't think we can verify nonces for receipts. @evgenykuzyakov maybe we should have hash instead?

Reply:
We can't remove nonce from the receipt, because the receipt will not be unique without a unique nonce. But we also shouldn't add hash, because it would be an additional data that flies through the network, that still needs to be recomputed to validate it.